### PR TITLE
Create region sensor

### PR DESCRIPTION
**Pull request by ghost**
Created region sensors, fixed some defects in HashSet & UnorderedArrayList, plus some cleanup

### CODE REVIEW

File: src/Utilities/HashSet.java [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
 /**
  * documentation
  *
  * @author davidrusu
  */
-public class HashSet {
+public class HashSet<T> {
```
Comment (davidrusu) [inline src/Utilities/HashSet.java:11]: 👍

File: src/gameengine/collisiondetection/tree/Tree.java [hunk 2]
```diff
@@ -133,17 +134,20 @@ public void entityUpdated(double timeToCheck, Entity entity) {
...
-            Shape.collideShapes(a.getShape(), b.getShape(), timeToCheck, temp);
-            if (temp.getCollisionTime() < result.getCollisionTime() - timeInTree) {
-                assert temp.getCollisionTime() <= timeToCheck : "too long" + temp
-                        .getCollisionTime() + ", " + timeToCheck;
-                result.set(temp);
-                result.setCollisionTime(result.getCollisionTime() + timeInTree);
-            }
+        if ((world.getCollisionGroups()[a.getEntityType()] & b.getEntityTypeBitMask()) == 0 || a
+                instanceof RegionSensor && ((RegionSensor) a).containsEntity(b) || b instanceof
+                RegionSensor && ((RegionSensor) b).containsEntity(a)) {
```
Comment (davidrusu) [inline src/gameengine/collisiondetection/tree/Tree.java:139]: What happens if both `a` and `b` are both region sensors?

File: src/gameengine/context/Context.java [hunk 3]
```diff
@@ -259,7 +260,7 @@ private RepeatedAction(long repeatDelay, boolean zeroDelayToFirstAction, boolean
                 freezeDelayOnPause, Action action) {
             assert repeatDelay > 0;
             this.action = action;
-            delay = repeatDelay * 1000;
+            delay = repeatDelay * GameTimer.NANOS_PER_MILLI;
```
Comment (davidrusu) [inline src/gameengine/context/Context.java:263]: 👍

File: src/gameengine/input/MouseController.java [hunk 4]
```diff
@@ -91,17 +91,11 @@ public MouseController(GameCore core, ScreenManager screen, int desiredFramerate
         }
     }
 
-    /**
-     * @see MouseProperties#isInvertedVerticalAxis()
-     */
```
Comment (davidrusu) [inline src/gameengine/input/MouseController.java:96]: 👍

File: src/gameengine/physics/Physics.java [hunk 5]
```diff
@@ -20,7 +20,10 @@ public static double getCriticallyDampedSpringConstant(double k, double mass) {
     }
 
     public static double springForce(Vector2D a, Vector2D b, double k, double restingLength) {
-        double displacement = new Vector2D(a).subtract(b).length() - restingLength;
+        double dx = a.getX() - b.getX();
+        double dy = a.getY() - b.getY();
+        double length = Math.sqrt(dx * dx + dy * dy);
+        double displacement = length - restingLength;
```
Comment (davidrusu) [inline src/gameengine/physics/Physics.java:26]: 😢

**Review by davidrusu — APPROVED** [review]

File: src/gameengine/collisiondetection/tree/Tree.java (hunk 2 above)
Comment (ghost) [reply src/gameengine/collisiondetection/tree/Tree.java:139]: So if the first condition (& bitmask) == 0 then we know that a and b don't interact and return right away.  If we keep checking the condition that means that the first condition is false so they should interact which means that the region sensors want to be aware of each other.  If one of the sensors already contains the other then they are already aware.

But... This does point out a defect in a different part of the code where I add an entity to the region sensor as I should add each other to the other (whereas right now only 1 region sensor is  notified).

**Pull request merged**